scripts/clean-shrinkwrap.js

```diff
@@ -15,29 +15,222 @@
+  } catch (error) {
+    if (error.code === 'MODULE_NOT_FOUND') {
+      return null;
+    }
+
+    throw error;
+  }
+};
+
+/**
+ * @summary Get the top level dependencies of a shrinkwrap object
+ * @function
+ * @private
+ *
+ * @param {String[]} shrinkwrapPath - path to shrinkwrap dependency
+ * @returns {Object} top level dependencies
+ *
+ * @example
+ * const dependencies = getTopLevelDependenciesFromShrinkwrapObject([ 'debug' ]);
+ */
+const getTopLevelDependenciesFromShrinkwrapObject = (shrinkwrapPath) => {
+  return _.get(getShrinkwrapDependencyManifest(shrinkwrapPath), [ 'dependencies' ], {});
+};
+
+/**
+ * @summary Get the dependency tree of a shrinkwrap dependency
+ * @function
+ * @private
+ *
+ * @param {Object} shrinkwrap - the shrinkwrap file contents
+ * @param {String[]} shrinkwrapPath - path to shrinkwrap dependency
+ * @returns {Object[]} dependency tree
+ *
+ * @example
+ * const dependencyTree = getDependencyTree(require('./npm-shrinkwrap.json'), [ 'drivelist' ]);
+ */
+const getDependencyTree = (shrinkwrap, shrinkwrapPath) => {
+  const dependencies = getTopLevelDependenciesFromShrinkwrapObject(shrinkwrapPath);
+
+  if (_.isEmpty(dependencies)) {
+    return [];
+  }
+
+  const object = getShrinkwrapDependencyObject(shrinkwrap, shrinkwrapPath);
+  const result = _.map(dependencies, (version, name) => {
+    const dependencyPath = _.has(object.dependencies, name) ? _.concat(shrinkwrapPath, [ name ]) : [ name ];
+    return getPrettyShrinkwrapDependencyObject(shrinkwrap, dependencyPath);
   });
+
+  return _.concat(result, _.flatMapDeep(result, (dependency) => {
+    return getDependencyTree(shrinkwrap, dependency.path);
+  }));
+};
+
+/**
+ * @summary Remove certain dependencies from a shrinkwrap object
+ * @function
+ * @private
+ *
+ * @description
+ * A shrinkwrap object is a recursive data structure, that apart
+ * from some extra metadata, has the following structure:
+ *
+ * {
+ *   ...
+ *   "dependencies": {
+ *     "<dependency_name>": <recursive definition>,
+ *     "<dependency_name>": <recursive definition>,
+ *     "<dependency_name>": <recursive definition>,
+ *     ...
+ *   }
+ * }
+ *
+ * The purpose of this function is to remove certain dependencies
+ * that match a blacklist. In order to do so, we start from the top
+ * level object, remove the blacklisted dependencies, and recurse
+ * if possible.
+ *
+ * @param {Object} shrinkwrap - the shrinkwrap file contents
+ * @param {String[]} blacklist - dependency blacklist
+ * @param {Object} qualifiers - qualifiers
+ * @param {Boolean} qualifiers.development - development dependency
+ * @param {Boolean} qualifiers.optional - optional dependency
+ * @returns {Object} filtered shrinkwrap object
+ *
+ * @example
+ * const shrinkwrapFile = require('./npm-shrinkwrap.json');
+ * const dependencyTree = getDependencyTree(shrinkwrapFile, [ 'drivelist' ]);
+ * const filteredShrinkwrap = removeDependencies(shrinkwrapFile, dependencyTree, {
```

Comment:
I think this example is wrong, because `getDependencyTree` returns `Object[]` but the `blacklist` parameter here is `String[]` ?